source/system_timer.c

```diff
@@ -162,30 +154,133 @@ void timer_sys_event_free(arm_event_storage_t *event)
+        // Periodic - check due time of next launch
+        timer->launch_time += timer->period;
+        if (TICKS_BEFORE_OR_AT(timer->launch_time, timer_sys_ticks)) {
+            // next event is overdue - queue event now
+            eventOS_event_send_timer_allocated(&timer->event);
+        } else {
+            // add back to timer queue for the future
+            timer_sys_add(timer);
+        }
+    }
     platform_exit_critical();
 }
 
-uint32_t timer_get_runtime_ticks(void)  // only used in dev_stats_internal.c
+uint32_t eventOS_event_timer_ticks(void)
 {
     uint32_t ret_val;
+    // Enter/exit critical is a bit clunky, but necessary on 16-bit platforms,
+    // which won't be able to do an atomic 32-bit read.
     platform_enter_critical();
-    ret_val = run_time_tick_ticks;
+    ret_val = timer_sys_ticks;
     platform_exit_critical();
     return ret_val;
 }
 
-int8_t eventOS_event_timer_request(uint8_t snmessage, uint8_t event_type, int8_t tasklet_id, uint32_t time)
+/* Called internally with lock held */
+static void timer_sys_add(sys_timer_struct_s *timer)
 {
-    platform_enter_critical();
+    uint32_t at = timer->launch_time;
+
+    // Find first timer scheduled to run after us, and insert before it.
+    // (This means timers scheduled for same time run in order of request)
+    ns_list_foreach(sys_timer_struct_s, t, &system_timer_list) {
+        if (TICKS_BEFORE(at, t->launch_time)) {
+            ns_list_add_before(&system_timer_list, t, timer);
+            return;
+        }
+    }
+
+    // Didn't insert before another timer, so must be last.
+    ns_list_add_to_end(&system_timer_list, timer);
+}
+
+/* Called internally with lock held */
+static int_fast8_t eventOS_event_timer_request_at_(const arm_event_t *event, uint32_t at, uint32_t period)
+{
+    if (TICKS_BEFORE_OR_AT(at, timer_sys_ticks)) {
+        return eventOS_event_send(event);
+    }
 
     // Because we use user-allocated events, they must get delivered to avoid
     // a leak. Previously this call queued timers for invalid tasks, then they
     // would go undelivered. Now it returns an error.
-    if (!event_tasklet_handler_id_valid(tasklet_id)) {
-        goto error;
+    if (!event_tasklet_handler_id_valid(event->receiver)) {
+        return -1;
+    }
+
+    sys_timer_struct_s *timer = timer_struct_get();
+    if (!timer) {
+        return -1;
+    }
+
+    timer->event.data = *event;
+    timer->launch_time = at;
+    timer->period = period;
+
+    timer_sys_add(timer);
+
+    return 0;
+}
+
+int_fast8_t eventOS_event_timer_request_at(const arm_event_t *event, uint32_t at)
+{
+    platform_enter_critical();
+
+    int_fast8_t ret = eventOS_event_timer_request_at_(event, at, 0);
+
+    platform_exit_critical();
+
+    return ret;
+}
+
+int_fast8_t eventOS_event_timer_request_in(const arm_event_t *event, uint32_t in)
+{
+    platform_enter_critical();
+
+    int_fast8_t ret = eventOS_event_timer_request_at_(event, timer_sys_ticks + in, 0);
```

Comment:
If it merely overflows, with a sensible value of `in`you get a wrapped timestamp (in the next 2^32 epoch), and the BEFORE_OR_AT() test correctly handles that wrap of timestamp.

But if the `in` value was >= 0x80000000 (regardless of current time), it would interpret that as the past and it would trigger immediately.

Arguably, maybe the in, after and every parameters should be signed?